titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/admission/AdmissionSanitizer.java

```diff
@@ -18,12 +18,34 @@
 
 import reactor.core.publisher.Mono;
 
-public interface AdmissionSanitizer<T> {
+/**
+ * Sanitizers have a MapReduce-like interface, and are executed concurrently in two phases:
+ * <p>
+ * 1. <tt>sanitize(input)</tt> is called on all registered {@link AdmissionSanitizer sanitizers} concurrently. The input
+ * parameter may or may not have passed through other sanitizers, no assumptions on what data has been already changed
+ * during the sanitization process can be made. It is safer to assume no other sanitizers will touch the same data,
+ * and the input is the original value sent to the system.
+ * <p>
+ * 2. <tt>apply(input, changes)</tt> is chained through all sanitizers sequentially. Each sanitizer adds its partial
+ * update to the entity modified by its predecessor in the chain. The final result includes all partial updates from all
+ * sanitizers. No assumptions should be made on the order in which sanitizers will be called.
+ *
+ * @param <T> entity being sanitized
+ * @param <S> partial (sanitized) data to be merged back to the entity in the second step
+ */
+public interface AdmissionSanitizer<T, S> {
+    /**
+     * Async method that produces clean and missing data elements to an entity.
+     *
+     * @return Returns a Mono/Single with all necessary changes that need to be applied later. It may be a
+     * {@link Mono#empty()} when no changes are to be made, or a {@link Mono#error(Throwable)} when the entity is invalid
+     */
+    Mono<S> sanitize(T entity);
+
     /**
-     * Async method to clean and add missing data elements to an entity
+     * Apply a partial update into the entity being sanitized.
      *
-     * @return Returns a Mono/Single that emits a cleaned up version. If no changes were made, the emitted item will
-     * share a reference with the source parameter.
+     * @return the modified entity with updates applied
      */
-    Mono<T> sanitize(T entity);
+    T apply(T entity, S update);
```

Comment:
Good idea on `Mono<Function<T, T>>`, I like it. I will make the change.

On the non-overlapping partial updates, you are right. I tried to convey that on the javadoc for the interface, let me know if it can be made more clear.